---
title: "Journal"
author: "Noud Klaassen"
output: html_document
---

[**Project 1: SuRE: Somatic Mutations K562**](#p1)  

* [Week 1 (15-08)](#w1)  
* [Week 2 (26-08)](#w2)  
* [Week 3 (02-09)](#w3)  
* [Week 4 (09-09)](#w4)  
* [Week 5 (16-09)](#w5)  
* [Week 6 (23-09)](#w6)  

[**Project 2: SuRE: Germline Indels**](#p2)

# Project 1: SuRE: Somatic Mutations K562 {#p1}

## Week 1{#w1}
#### 15-08-2019 (Th)
VCF-files of K562 have been loaded in RStudio (non-phased) via the `wget` command in Linux. Exploratory analyses on the nr. of SNPs per chromosome, average QUAL-scores.

#### 16-08-2019 (Fr)
Received feedback from Ludo on the `RMarkdown`-file that was created. Some more exploratory analysese were performed

## Week 2{#w2}
#### 26-08-2019 (Mo)
I generated a RProject for the K562_SuRE directory. I linked the files to the GitHub repository which I will commit every day at the end of the day. Also I started exploring the uniqueness in the MNVs that I have found. It seems that 0 MNVs start at the same position, but 0.5% of MNVs have some overlap with other MNVs

#### 27-08-2019 (Tu)
I to find a way to identify overlapping Indels and SNVs. I made a function that was able to do this up to 50.000 rows at a time (which is not enough). Variants were annotated with the `TxDb.Hsapiens.UCSC.hg19.knownGene`-package (supplying the transcript database) and the `VariantTools`-package (allowing for variant identification)

#### 28-08-2019 (We)
Exploratory analysis to identify SNVs, new vs known variants without the use of extra packages. Talked with Bas & Ludo about the project. Need to filter heterozygotes and generate a distributionplot of the location of mutations on a chromosome

#### 29-08-2019 (Th)
I rewrote the whole `20190823_VariantCalling.R`-script. Primarily all information (e.g. new vs known variant) was put into a new `GRanges` object. With the new script, all this information is stored within the metadata of the same `GRanges` object. The desired objects can therefore easily be filtered out by using boolean arguments. Also started on the visualization of data using a distribution plot. For this I used the `karyoploteR` function. 

#### 30-08-2019 (Fr)
Tried to annotate transcription factor binding domains. Didn't work because the `MotIV`-package required GSL which was hard to install and link properly. Installed conda on the server. Got the dotplot function of the `KaryoplotR` working.

## Week 3{#w3}
#### 02-09-2019 (Mo)
Tried some more `KaryoplotR` functions. I wanted to annotate the genomic positions with Transcription Factor Binding Sites (TFBS). Therefore i downloaded the `RegulomeDB.dbSNP138`. However, this database could not be loaded in R. Also, it only contains information on SNPs present in the dbSNP.

#### 03-09-2019 (Tu)
Tried to find a way to count the total number of exons in the genes etc.

#### 04-09-2019 (We)
I found a way to remove duplicates from the Variantannotation. In this way, 104% of variants is represented, indicating still some overlap between various variants. This could e.g. be a promoter within an intergenic region etc. Also started exploring heterozygotic mutations in regions with Loss of Heterozygosity (LOH). Strangely, there are still a lot of heterozygotic mutations in these regions. 

#### 05-09-2019 (Th)
It turns out that the VCF-file from the K562-paper doesn't correctly annotate the genotype. In regions that have loss of heterozygosity (e.g. chromosome 13 and 14), the genotype is often described as 1/1 (homozygote for the alternative allele), while there would be only 1 allele. This is a problem as we are mostly interested in heterozygous mutations. Regions with CN = 1 should therefore be discarded. 

#### 06-09-2019 (Fr)
I compared phased and non-phased data from the K562-paper. There are minimal differences, we choose to continue with the phased data. Wrote a script to generate the right vcf-files and the right txt files. Successfully generated these files with the `20190906_PreparingVCF` script
```{r, echo = F}
library(knitr)
Files = c("VCF files", "Txt files (All)", "Txt files (New)")
Type = c(".vcf.gz", ".txt",".txt")
Variants = c("All", "All", "New")
Overlap = c("No", "No", "No")
Directory = c("~/projects/SuRE_K562/data/interim/SNPs/vcf", "~/projects/SuRE_K562/data/interim/SNPs/txt", "~/projects/SuRE_K562/data/interim/SNPs/txt")
df = data.frame(Files, Type, Variants, Overlap, Directory)
kable(df)
```

### Week 4{#w4}

#### 09-09-2019 (Mo)
I downloaded `bed_files.tar.gz` from `http://jaspar.genereg.net/download/bed_files.tar.gz` to function as a database for transcription factor binding sites. I previously did the same with `RegulomeDB.dbSNP132.Category3.txt` from `http://www.regulomedb.org/downloads/RegulomeDB.dbSNP132.Category3.txt.gz`, but this file could not be loaded into R. `MA1198.1.bed` was removed by using the MobaXterm as it contains no data. 

#### 10-09-2019 (Tu)
I got the bedfiles annotation of the TFBS working for the `GRanges` file. Able to generate a `GRanges` file with all novel mutations and the corresonding number of the file of the TFBS they interfere with. Mind that mutations that alter multiple TFBS occur in multiple rows. For the SuRE-pipeline, I resolved a few errors regarding the SNPs database annotation. We use the non-phased files. Important to go to the conda environment of K562_SuRE, otherwise, python is not correctly installed (`conda activate K562_SuRE`)

#### We-Fr
Prepare the SuRE pipeline and files for the sure pipeline.

### Week 5{#w5}

#### 16-09-2019 (Mo)
Analyse data for the SuRE pipeline with the `20190913_SuRE_Explore.R` script.

#### 17-09-2019 (Tu)
Analyse data for the SuRE pipeline

#### 18-09-2019 (We)
Analyse data for the SuRE pipeline with a more clean script `20190916_SuRE_Count_Import.R`

#### 19-09-2019 (Th)
Figured out that the data is really not of that good and sufficient quality. Of the 3 million SNVs only 9,176 have >10 cDNA reads per variant location and are also novel. Some concern arrises on whether mutations in dbSNP could be somatic. 

#### 20-09-2019 (Fr)
Analysed the SuRE data some more and figured out that for all novel SNP there are only two significant hits. This is a problem and we most likely do not continue with the analysis of this dataset. 

### Week 6{#w6}

#### 23-09-2019 (Mo)
The possibility that somatic mutations are present in dbSNP was explored a little further. Within literature many different practices are used. Some remove all dbSNP hits. Others are working with logg odss ratios (LOR) and dbSNP and COSMIC (somatic mutations in cancer database). Some publicatios make use of the `MuTect` software that also employs the dbSNP to differentiate probabilities of calls. 

#### 24-09-2019 (Tu)
Literature research on the COSMIC and dbSNP databases. It is still not clear to what extent somatic mutations are present in dbSNP. However, from the known SNPs in the K562 paper there is 4.8% (145,792 variants) that are present in the COSMIC database. 

#### 25-09-2019 (We)
Discussed again with Bas and Ludo that we are going to leave the K562 paper behind. Now we are going to work on the indels. 

# Project 2: SuRE: Germline Indels {#p2}

#### 18-10-2019 (Fr)
Last week I have worked a lot on the data for the indels. First there was a problem with the "SNP_PARENT" column, that was not annotated correctly for the indels. Then, using the script that normalizes the data and saves the data in one (1) variant per row, it went wrong with the SuRE44_2 repeat. It turned out that this error was due to the fact that some columns in "SNP_ABS_POS_hg19"