---
title: "20190906_PreparingVCF"
author: "Noud Klaassen"
date: "6-9-2019"
output: html_document
---

### Startup

Data generated by the K562 paper (doi: 10.1101/gr.234948.118) was downloaded from `https://www.encodeproject.org/files/ENCFF241LKI/@@download/ENCFF241LKI.vcf.gz` using `wget 'https://www.encodeproject.org/files/ENCFF241LKI/@@download/ENCFF241LKI.vcf.gz'`. This datafile is ~3 GB and contains variant calls from the phased sequencing data. Required libraries were loaded. The data was read to a `CollapsedVCF` object with the `readVcf` function. To be able to use subsets of data from this object, a `TabixFile` was generated.  

```{r, eval = FALSE}
library(VariantTools)
library(vcfR)
library(R.utils)
fl <- "~/projects/SuRE_K562/data/external/Encode_K562_VCF/ENCFF241LKI.vcf.gz"
vcf <- readVcf(fl, genome = "hg19")
compressVcf <- bgzip(fl, tempfile())
idx <- indexTabix(compressVcf, "vcf")
tab <- TabixFile(compressVcf, idx)
```

### Write to .vcf.gz

With this function, i subset the vcf per chromosome to generate a new vcf file. This file is consequently gzipped to .vcf.gz format. In this function, a loop is used over all chromosomes (1 to 22 and X). `GRanges` for these chromosomes are determined and put into the `param` object. The VCF-file can be read in again. Here, the `tab` object us used to refer to the VCF-file. 

```{r, eval = FALSE}
for (chrom in paste0("chr", c(1:22, "X"))){
    
  ranges <- as(seqinfo(vcf)[chrom], "GRanges")
  param <- ScanVcfParam(which = ranges)
  vcf.chr <- readVcf(tab, genome = "hg19", param = param)
  dir = paste0("~/projects/SuRE_K562/data/interim/SNPs/vcf/All_K562_Phased_", chrom, ".vcf")
  writeVcf(vcf.chr, dir)
  print(paste0("All_K562_Phased_", chrom, ".vcf", " was successfully generated"))
    
  }
```

The obtained `*.vcf` files can be gzipped with the `gzip` command. Again, a loop is used to do this for all the files. 

```{r, eval = FALSE}
for (chrom in paste0("chr", c(1:22, "X"))){
    dir = paste0("~/projects/SuRE_K562/data/interim/SNPs/vcf/All_K562_Phased_", chrom, ".vcf")
    gzip(dir, remove = FALSE)
    print(paste0("All_K562_Phased_", chrom, ".vcf", " was successfully gzipped to", " All_K562_Phased_", chrom, ".vcf.gz"))
  }
```

### Write to .txt (All variants)

Besides a `.vcf` file, also a `.txt` file is needed as a reference object. These textfiles (per chromosome) should contain startposition | reference | alternative in a tab-delimited fashion. `GRanges` are obtained from the `vcf` object and overlapping variants (e.g. SNP within deletion) are removed. `gr_chrom` contains the genomic regions for that specific chromosome. startpositions can be obtained with the `start` function over the `IRanges` object. Reference sequences can be easilily obtained by calling the `$REF` column. For the alternative sequences, this is not possible as these can contain multiple sequences. This is resolved with the `CharacterList` and `unstrsplit` functions. A dataframe is generated which is subsequently written to a `.txt` file.  

```{r,eval=FALSE}
gr <- granges(vcf)
gr_no.overlap <- gr[countOverlaps(gr) == 1]
    
# start the loop here
    
for (chrom in paste0("chr", c(1:22, "X"))){
  gr_chrom <- gr_no.overlap[seqnames(gr_no.overlap) == chrom]
  start <- start(ranges(gr_chrom))
  ref <- as.character(gr_chrom$REF)
  altlist <- CharacterList(gr_chrom$ALT)
  alt <- unstrsplit(altlist, sep=",")
  df <- data.frame(start, ref, alt, stringsAsFactors = FALSE)
  dir = paste0("~/projects/SuRE_K562/data/interim/SNPs/txt/All_NoOverlap_K562_Phased_", chrom, ".txt")
  write.table(df, file = dir, row.names = FALSE, col.names = FALSE, quote = FALSE, sep = "\t")
  print(paste0("All_NoOverlap_K562_Phased_",chrom,".txt", " was successfully generated"))
}  
```

### Write to .txt (New variants)

For the new variant, the same workflow was used as for all variants. Again, only non-overlapping variants are being used. A new `GRanges` object is created that only contains the new variants. This is done with the `grepl` function which returns `TRUE` values when the name contains an rs-number. This is reverted with the `!`. So only novel variants (not in dbSNP138; no rs-number). 

```{r, eval=F}
gr_no.overlap_new <- gr_no.overlap[!grepl("rs", names(gr_no.overlap))]
    
    for (chrom in paste0("chr", c(1:22, "X"))){
      gr_chrom <- gr_no.overlap_new[seqnames(gr_no.overlap_new) == chrom]
      start <- start(ranges(gr_chrom))
      ref <- as.character(gr_chrom$REF)
      altlist <- CharacterList(gr_chrom$ALT)
      alt <- unstrsplit(altlist, sep=",")
      df <- data.frame(start, ref, alt, stringsAsFactors = FALSE)
      dir = paste0("~/projects/SuRE_K562/data/interim/SNPs/txt/New_NoOverlap_K562_Phased_", chrom, ".txt")
      write.table(df, file = dir, row.names = FALSE, col.names = FALSE, quote = FALSE, sep = "\t")
      print(paste0("New_NoOverlap_K562_Phased_",chrom,".txt", " was successfully generated"))
    }
```