---
title: "20190909_TFBS_Annotation"
author: "Noud Klaassen"
date: "10-9-2019"
output: html_document
---

```{r setup, echo=F, message = FALSE}
library(vcfR)
library(VariantTools)
```

First the data is loaded. Data is a Phased VCF file downloaded from `https://www.encodeproject.org/files/ENCFF241LKI/@@download/ENCFF606RIC.vcf.gz` which was retrieved by using `wget 'https://www.encodeproject.org/files/ENCFF241LKI/@@download/ENCFF606RIC.vcf.gz'` on 2019-08-20. `.bed` files were downloaded from `http://jaspar.genereg.net/download/bed_files.tar.gz` using `wget 'http://jaspar.genereg.net/download/bed_files.tar.gz'`, which contain genomic coordinates of the Transcription Factor Binding Site (TFBS) sequences. 
```{r loading_data}
vcfK <- readVcf("~/projects/SuRE_K562/data/external/Encode_K562_VCF/ENCFF606RIC.vcf.gz", genome = "hg19")
gr <- rowRanges(vcfK)
files <- list.files("~/projects/SuRE_K562/data/external/bed_files/bed")
```

The next loop puts all data from the `.bed` files in the `out` dataframe. With the `if` loop, I select only the bedfiles that contain human TFBS. This `out` dataframe contains (in rising column order): chromosome, start position of the TFBS (0-based), end position of the TFBS (0-based), index including corresponding chromosome, empty column, strand (plus or minus), corresponding file. 
```{r}
out <- NULL
for (file in files){
  tab <- read.table(paste0("~/projects/SuRE_K562/data/external/bed_files/bed/", file, sep=""), stringsAsFactors = FALSE)
  tab$File <- rep(file, nrow(tab))
  if (grepl("hg19", tab$V4[1]) == FALSE){
    next
  }
  out <- rbind(out,tab)
}
out[1,]
```

Then a `GRanges` files is created from the `dataframe`. Note that the start-position is increased with 1 with respect to the bed-file, to change from a 0-based to 1-based system. Ranges are sorted based on their chromosomal position. 
```{r}
SEQ   <- out$V1
START <- out$V2+1 #0-based to 1-based
END   <- out$V3

gr_tfbs <- GRanges(seqnames = SEQ, 
                   ranges = IRanges(start = START, end = END), 
                   File = out$File, 
                   strand = out$V6)  
gr_tfbs <- sortSeqlevels(gr_tfbs)
gr_tfbs <- sort(gr_tfbs)
OVERLAP <- gr[countOverlaps(gr, gr_tfbs) > 0]
table(countOverlaps(gr, gr_tfbs))

```

Then the original file can be annotated with the corespondig files
```{r}
gr_new <- gr[gr$NEW == T]
f <- findOverlaps(gr_new, gr_tfbs)
q <- queryHits(f)
s <- subjectHits(f)

gr_tfbs$File[s] 

OVERLAP.REDUNDANT <- gr_new[q]
OVERLAP.REDUNDANT$FILE <- gr_tfbs$File[s]
```