---
title: "20190923_COSMIC_analysis"
author: "Noud Klaassen"
date: "25-9-2019"
output: html_document
---

This script is used to download data from COSMIC (Catalogue Of Somatic Mutations In Cancer) and compare this to various other data. In this RMarkdown file I use the `nrows = 200` argument, to speed up the knitting of this file. First the `data.table` and `GenomicRanges` packages are loaded for loading and transfer to `GRanges` object.

```{r library loading, message = F}
library(data.table)
library(GenomicRanges)
```

## Data Downloading & Reformatting
Data is a tab separated value file (`.tsv.gz`) which was downloaded from [https://cancer.sanger.ac.uk/cosmic/download](https://cancer.sanger.ac.uk/cosmic/download) (v90 release 5th september 2019) on 25-09-2019 on the Windows machine and transferred to the RHPC server via WinSCP. Login is necessary for the download. Next script is used to read the files into R and reformat them for further analysis
```{r data download and reformat, message = F}
dir = "/DATA/usr/n.klaassen/projects/SuRE_K562/data/external/COSMIC/"
cosmic.noncoding <- fread(file = paste0(dir, "CosmicNCV.tsv.gz"), nrows = 10000)
colnames(cosmic.noncoding)

# Remove columns that are unnessesary
cosmic.noncoding.filter.rows <- cosmic.noncoding
cosmic.noncoding.filter.rows[,c(1,3:11,13,21:28)] <- NULL
cosmic.noncoding.filter.rows[11:16]
```

## Filter variants
This script is used to remove al variants that are not SNVs (e.g. indels) and annotate genomic positions. This is done by requiring both the length of the WT sequence and the MUT sequence to be equal to 1. 
```{r keep SNV}
# Remove Insertions and deletions
filter.SNV <- nchar(cosmic.noncoding.filter.rows$WT_SEQ) == 1 & nchar(cosmic.noncoding.filter.rows$MUT_SEQ) == 1
cosmic.noncoding.filter.rows.SNV <- cosmic.noncoding.filter.rows[filter.SNV]
cosmic.noncoding.filter.rows.SNV[11:16]
table(cosmic.noncoding.filter.rows.SNV$SNP)

# Generate genomic positions
genomic.locations <- paste0("chr", cosmic.noncoding.filter.rows.SNV$`genome position`)
cosmic.granges <- as(genomic.locations, "GRanges")
seqlevels(cosmic.granges)
```
As can be seen, the `cosmic.granges` object contains the seqlevels chr1 to chr25. Probably the chr23 - chr25 corresponds to chrX, chrY and chrM. However, I am not sure of this. To avoid problems, only the variants within chromosomes 1 till 22 were selected, sorted corresponding to their chromosome number and saved
```{r granges reformat}
cosmic.granges.chrom <- cosmic.granges[seqnames(cosmic.granges) %in% paste0("chr", c(1:22)),]
cosmic.granges.chrom <- sort(sortSeqlevels(cosmic.granges.chrom))
```

## Compare with K562 paper variants
To compare the COSMIC somatic variants with the ones in the K562 paper, this data (unphased) was read into R. The interesting selection of variants are SNVs (no indels or structural variants) and are already in the dbSNP (and therefore not used in previous analyses). These variants are saved into the `gr.rs.SNV` object.

```{r read K562, eval = F}
library(vcfR)
vcfK <- readVcf("~/projects/SuRE_K562/data/external/Encode_K562_VCF/ENCFF606RIC.vcf.gz", 
                genome = "hg19")
gr <- rowRanges(vcfK)
gr$SNV <- isSNV(vcfK, singleAltOnly = FALSE)
gr$NEW <- !grepl("rs", names(gr))
gr.rs.SNV <- gr[gr$SNV == T & gr$NEW == F]
```

Then the overlap between the known K562 variants and the COSMIC can be determined with the `countOverlaps` function. This function determines, for the known variants in the K562 paper (`gr.rs.SNV`) with how many SNVs of the COSMIC database (`cosmic.granges.chrom`) they overlap. With the `sum` function it can be determined how many variants actually overlap. 
```{r countoverlaps, eval = F}
overlap <- countOverlaps(gr.rs.SNV, cosmic.granges.chrom)
tab <- table(overlap)
sum(tab[2:length(tab)])
cosmic.overlap.snp.id <- names(overlap[overlap > 0])
```


