# This script is designed to process generated by the SuRE-Pipeline
# to get a general overview of the data quality. 

# First load the required libraries
# NO LIBRARIES NEEDED YET
library(data.table)
library(tidyverse)

#### Loading the SNPtxt files into R ####
# This can be done with the txt.gz files directly and do not need 
# to be unzipped

dir <- "/DATA/usr/n.klaassen/projects/SuRE_K562/data/processed/SuRE_output/SuRE-counts_"

for (chrom in paste0("chr", c(1:22, "X"))){ #THIS FUNCTION TAKES 30 MIN TO COMPLETE!!!!
  #READ TXT.GZ
    tab <- fread(paste0(dir, chrom, ".txt.gz"), header = TRUE, sep = "\t", stringsAsFactors = FALSE)
    print(paste0(chrom, " is read at", Sys.time()))
  
  #SEPARATE ROWS
    name <- paste0("sure", chrom)
    tabsep <- separate_rows(tab, SNPrelpos:SNPidx, SNPbaseInf:PAT_MAT, sep = ",") 
    assign(name, tabsep)
    print(paste0(chrom, " is separated at", Sys.time()))
  
  #WRITE SEPARATE TXT FILES
    write.table(get(paste0("sure", chrom)), paste0("/DATA/usr/n.klaassen/projects/SuRE_K562/data/processed/SuRE_Counts_Separated_20190917/SuRE-Counts_Separated_", chrom,  ".txt"), 
            quote = FALSE, 
            sep = "\t", 
            row.names = FALSE, 
            col.names = TRUE)
    print(paste0(chrom, "is writen to .txt file at ", Sys.time()))
}

####  READ SEPARATED FILES ####

for (chrom in paste0("chr", c(1:22, "X"))){
  assign(paste0("sure", chrom), fread(paste0("/DATA/usr/n.klaassen/projects/SuRE_K562/data/processed/SuRE_Counts_Separated_20190917/SuRE-Counts_Separated_", chrom,  ".txt"), 
             header = TRUE, 
             sep = "\t", 
             stringsAsFactors = FALSE,
             na.strings = NULL))
}

#### CONCATONATE DATATABLES ####

suresep = NULL
for (chrom in paste0("chr", c(1:22, "X"))){
  suresep <- rbind(suresep, get(paste0("sure", chrom)))
  print(paste0(chrom, "is added to datatable"))
}

surechr1sub <- na.omit(surechr1)
suresep <- surechr1sub



############################################
#### Summarize biological replicates #######
############################################

suresep$SuRE_B1 <- suresep$SuRE23_45_B1_T1+suresep$SuRE23_45_B1_T2+suresep$SuRE23_45_B1_T3+suresep$SuRE23_45_B1_T4
suresep$SuRE_B2 <- suresep$SuRE23_55_B2_T1+suresep$SuRE23_55_B2_T2
suresep$SuRE_B  <- suresep$SuRE_B1+suresep$SuRE_B2
suresepsub <- suresep[grepl("chr", suresep$SNP_ID) & suresep$SuRE_B > 10 & suresep$PAT_MAT != 3]

##################################################
#### General Statistics ##########################
##################################################

sureuni <- sure[!grepl(",", sure$SNPbase),]
total <- nrow(sure) #Total amount of observations
nonuni <- (nrow(sure)-nrow(sureuni)) #amount of rows contain multiple variants
uniknown <- sum(grepl("rs", sureuni$SNP_ID)) #amount known variants
uniunknown <- sum(grepl("chr", sureuni$SNP_ID)) #amount unknown variants

countpervariant <- as.data.frame(tapply(suresep$SuRE_B, suresep$SNP_ID, function(x){sum(x, na.rm = TRUE)})) #SuRE-counts per SNP_ID; first row should be deleted
countpervariant <- data.frame(countpervariant[-1,], row.names = rownames(countpervariant)[-1])
colnames(countpervariant) <- "FREQ"
boxplot(countpervariant$FREQ,col = 2, ylab = "Transcript reads (B1 + B2)")
summary(countpervariant$FREQ)

sum(grepl("chr", rownames(countpervariant))) #Amount of unique novel variants
sum(countpervariant$FREQ > 0) # amount of variants with cDNA counts higher than 10
sum(countpervariant$FREQ > 0 & grepl("rs", rownames(countpervariant))) #amount of novel variants with cDNA counts higher than 10



