---
title: "20190823_VariantCalling"
author: "Noud Klaassen"
date: "23-8-2019"
output: html_document
---

Load libraries; Data is a VCF file downloaded from `https://www.encodeproject.org/files/ENCFF606RIC/@@download/ENCFF606RIC.vcf.gz` which was retrieved by using `wget https://www.encodeproject.org/files/ENCFF606RIC/@@download/ENCFF606RIC.vcf.gz'` on 2019-08-20
```{r Library loading, message = FALSE}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(VariantTools)
library(VariantAnnotation)
library(karyoploteR)
library(ggplot2)
library(vcfR)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
vcfK <- readVcf("~/projects/SuRE_K562/data/external/Encode_K562_VCF/ENCFF606RIC.vcf.gz", 
               genome = "hg19")
gr <- rowRanges(vcfK)
gr$GT <- geno(vcfK)$GT #Add a metadatacolumn for genotype
```

An exploratory analysis was performed to disect different headers from the vcf-file. 
```{r Exploratory, eval = FALSE, echo=FALSE}
  fl <- system.file("extdata", "chr22.vcf.gz", package="VariantAnnotation")
  vcf <- readVcf(fl, "hg19")
  header(vcf) #shows the header
  samples(header(vcf)) #shows subset 'samples' of the header
  geno(header(vcf)) #shows subset 'geno' of the header
  head(rowRanges(vcf),5) #retrieve GRanges object of 3 rows
  ref(vcf)[1:5] #retrieve REF sequence for the first 5 rows in a DNAStringSet object
  qual(vcf)[1:5] #retrieve QUAL scores of the first 5 rows
  alt(vcf)[1:5] #retrieve ALT sequences (could be more than 1)
  geno(vcf) #shows al genotypes as described in FORMAT fields
  sapply(geno(vcf), class)
  info(vcf) #shows some sort of summary of all the data. Rows are indicated by rs-numbers or chromosome positions 
  rowRanges(vcf) #Gives the Granges of all rows (what is the difference with head(Rowranges)???)
  as.data.frame(rowRanges(vcf)) #generates dataframe of these rows
  vcf[1:100,] #subset few rows within a VCF
  seqlevels(vcf) <- "chr22" #set the seqlevels all to "chr22" instead of 22 (this is the required format)
  rd <- rowRanges(vcf)
  lengths(gr$ALT) #determine the amount of ALT sequences
  table(width(gr_snv)) #contingency table for the length of the DNA fragments (REF or ALT??), ik vermoed REF
  expand(vcf) #generates an "expanded-VCF" object that shows a row for every alternative sequence
```

It is important to note that all analyses here where done on the same `GRanges` file. This means that all filters apply for all 3,800,959 variants within the VCF. In this method, information on a certain variant (its genotype, SNV vs other variants or whether it is a new or existing variant) were annotated by adding metadata columns in the `GRanges`-file. Genotypes were obtained from the `CollapsedVCF`-object with the `geno()`-function. The `is_het()`-function from the `vcfR`-package is able to distinguish between heterozygotes and homozygotes on the basis of this genotype-matrix.SNVs (in contrast to indels and other structural variants) were identified with the `isSNV()`-function. This is the only function that only works on the CollapsedVCF-object and not on the GRanges-object. 'New' variants were defined as not having an rs-number. Overlapping variants (e.g. SNV within an Indel) were identified with the `countOverlaps()`-function.

This way, the desired `GRanges`-file can be created using boolean arguments like 
```length(gr[gr$HET == TRUE & gr$SNV == TRUE & gr$NEW == TRUE & gr$OVERLAP2 == FALSE])```
```{r Metadata generation}
gr <- rowRanges(vcfK) #generetes a GRanges object from the vcf-file
gr$GT <- geno(vcfK)$GT #Add a metadatacolumn for genotype (i.e. 0/0/1/1)
gr$HET <- is_het(gr$GT, na_is_false=TRUE) #Add a metadatacolumn for heterogeneity (TRUE = heterozygote, FALSE = homozygote)
gr$SNV <- isSNV(vcfK, singleAltOnly = FALSE) #Add a metadatacolumn for SNV (TRUE = SNV, FALSE = else)
gr$NEW <- grepl("chr", names(gr)) #Add a metadatacolumn for whether the mutation is new (TRUE) or already has an rs-number (FALSE)
gr$OVERLAP <- countOverlaps(gr) > 1
gr[1667]
```

```{r Bargraphs, echo=FALSE}
#Plot SNV vs other
b <- barplot(table(gr$SNV), 
             main = "SNVs",
             ylab = "Frequency",
             ylim = c(0,3500000),
             names.arg = c("Other (e.g. Indel)", "SNV"))
text(x=b, y=table(gr$SNV)+200000, labels = as.character(table(gr$SNV)))

#Plot nr. of Overlaps
b <- barplot(table(countOverlaps(gr)), 
             main = "Overlapping variants",
             xlab = "Number of overlaps +1",
             ylab = "Frequency",
             ylim = c(0,4200000),
                          )
text(x=b, y=table(countOverlaps(gr))+200000, labels = as.character(table(countOverlaps(gr))))

#Plot NEW vs rs
b <- barplot(table(gr$NEW), 
             main = "Novel variants (no rs-number)",
             ylab = "Frequency",
             ylim = c(0,4200000),
             names.arg = c("dbSNP138", "Novel"))
text(x=b, y=table(gr$NEW)+200000, labels = as.character(table(gr$NEW)))

#Plot Heterozygote vs Homozygote
b <- barplot(table(gr$HET), 
             main = "Variant heterozygosity",
             ylab = "Frequency",
             ylim = c(0,2500000),
             names.arg = c("Homozygote", "Heterozygote")
)
text(x=b, y=table(gr$HET)+200000, labels = as.character(table(gr$HET)))

#Plot summarizingdata 
grp <- gr[gr$OVERLAP == F]
data <- matrix(c(
               length(grp[grp$NEW == F]),
               length(grp[grp$NEW == T]),
               length(grp[grp$NEW == T & grp$SNV == F]),
               length(grp[grp$NEW == T & grp$SNV == T]),
               length(grp[grp$NEW == T & grp$SNV == T & grp$HET == T]),
               length(grp[grp$NEW == T & grp$SNV == T & grp$HET == F]))
               , nrow = 2) #First fills the columns, then the rows
colnames(data) <- c("All", "Novel", "Novel SNV")
data_percentage <- apply(data, 2, function(x){x*100/sum(x)})

b <- barplot(data_percentage, 
             main = "Variants (non-overlapping)",
             xlab = NULL,
             ylab = "Percentage",
             ylim = c(0,130),
             names.arg = colnames(data),
)
text(x=b, 
     y=c(data_percentage[1,])+3, 
     labels = data[2,], srt = 0,
     #adj = c(0.5,2)
     )
text(x=b, 
     y=c(data_percentage[1,])-3, 
     labels = data[1,], srt = 0,
     #adj = c(0.5,1.5),
     col = 0
     )
text(x=b-0.35,
     y=10,
     labels = c("dbSNP138", "Other (e.g. Indel)", "Heterozygote"),
     srt = 60,
     adj = c(0,0.5),
     col = 0)
text(x=b-0.35,
     y=10,
     labels = c("dbSNP138", "Other (e.g. Indel)", "Heterozygote"),
     srt = 60,
     adj = c(0,0.5),
     col = 0)
text(x=b,
     y=105,
     labels = c("Novel","SNV", "Homozygote"),
     srt = 60,
     adj = c(0,0.5))

```

Variants were annotated with the `VariantAnnotation` tool. Usefull info on `https://www.bioconductor.org/help/course-materials/2014/BioC2014/Lawrence_Tutorial.pdf`. The `locateVariants`-function is able to distinguish between different kinds of transcripts which are `spliceSite`, `intron`, `fiveUTR`, `threeUTR`, `coding`, `intergenic`, and `promoter`. This function requires a `GRanges` object, a link to a transcript annotation database (in this case hg19 UCSC) and the desired transcript. Unfortunately, a single variant will receive multiple `intron`-annotations, artificially increasing the prevalence of this variant annotation. 
```{r Variant annotation, message = FALSE, results="hide", warning=FALSE}
gr_all <- locateVariants(gr, txdb, AllVariants()) #locate the variants

gr_in <- locateVariants(gr, txdb,IntronVariants())  
gr_inu <- gr_in[!duplicated(gr_in$QUERYID)]
  
gr_ig <- locateVariants(gr, txdb,IntergenicVariants())
gr_igu <- gr_ig[!duplicated(gr_ig$QUERYID)] ## Unique_Intergenic (no values discarded)

gr_c <- locateVariants(gr, txdb, CodingVariants()) 
gr_cu <- gr_c[!duplicated(gr_c$QUERYID)] ## Unique_Coding (28589 values discarded)

gr_p <- locateVariants(gr, txdb, PromoterVariants())
gr_pu <- gr_p[!duplicated(gr_p$QUERYID)]

gr_3 <- locateVariants(gr, txdb, ThreeUTRVariants())
gr_3u <- gr_3[!duplicated(gr_3$QUERYID)]

gr_5 <- locateVariants(gr, txdb, FiveUTRVariants())
gr_5u <- gr_5[!duplicated(gr_5$QUERYID)]

gr_sp <- locateVariants(gr, txdb, SpliceSiteVariants())
gr_spu <- gr_sp[!duplicated(gr_sp$QUERYID)]

count <- c(length(gr),
           length(gr_inu),
           length(gr_igu), 
           length(gr_pu), 
           length(gr_cu),
           length(gr_3u),
           length(gr_5u),
           length(gr_spu))
percentage <-  c(count/count[1]*100)

var <- matrix(data=c(count, percentage), 
              ncol = 2, 
              dimnames = list(c("All", "Intron", "Intergenic", "Promoter", "Coding", "3-UTR", "5-UTR", "SpliceSite"),
                              c("Count", "Percentage")))

```
```{r Variant annotation plotting}
table(gr_all$LOCATION) #To check where the variants are ##doesnt work; way to many variants -> 3.9Mc
barplot(table(gr_all$LOCATION)/1000000,ylab="Frequency (*10^6)", las = 2)
var
```

To further explore the variant dataset, the density of the variants was assessed with the `karyoploteR` package. A helpfull website for this package was found on `https://bernatgel.github.io/karyoploter_tutorial//Tutorial/PlotParams/PlotParams.html`. The `ggplot`-package is used for the `alpha`-function, allowing for transparant colors. The `karyoploteR`-package allows for visualization of `GRanges`-objects on a Karyoplot. The variable `pp` is used temporarily to display the title closer to the plot. The next arguments are used to generate the plot for chromosome 1 only. Gray displays the densityplot of the `GRanges`-object containing all (3,9 million) variations while red only displays the novel (unannotated; 219,003) variations. It seems that the gray and red distributions follow a similar patterns (i.e. there are no regions that are enriched for novel mutations). Note that the paterns between red and gray should compared, not the actual values.

```{r Karyoplot}
pp <- getDefaultPlotParams(plot.type=1)
pp$topmargin <- 15
pp$leftmargin <- 0.2
kp <- plotKaryotype(plot.type = 1, 
                    chromosomes = "chr1", 
                    main = "Mutation density plot (including SNVs and other variants)",
                    plot.params = pp
                    )
kp <- kpDataBackground(kp)
kp <- kpAxis(kp, ymin = 0, ymax = 1, )
kp <- kpAddLabels(kp, labels = "All", col = "gray", r0 = 0.75, label.margin = 0.01, srt = 45, cex = 2)
kp <- kpAddLabels(kp, labels = "Novel", col = "red", r0 = 0.35, label.margin = 0.01, srt = 45, cex = 2)
kpPlotDensity(kp, data=gr, col = alpha("gray", 1), window.size = 1000000, border = 1)
kpPlotDensity(kp, data=gr[gr$NEW == TRUE], col = alpha("red", 0.6), window.size = 1000000, border = 1)
```
```{r Karyoplot2, echo=FALSE}
pp <- getDefaultPlotParams(plot.type=1)
pp$topmargin <- 15
pp$leftmargin <- 0.2
kp <- plotKaryotype(plot.type = 1, 
                    chromosomes = "chr1", 
                    main = "Mutation density plot (including SNVs and other variants)",
                    plot.params = pp
)
kp <- kpDataBackground(kp)
kp <- kpAxis(kp, ymin = 0, ymax = 1, )
kp <- kpAddLabels(kp, labels = "Intergenic", col = "gray", r0 = 0.75, label.margin = 0.01, srt = 45, cex = 2)
kp <- kpAddLabels(kp, labels = "Genes", col = "red", r0 = 0.35, label.margin = 0.01, srt = 45, cex = 2)
kpPlotDensity(kp, data=gr_ig, col = alpha("gray", 1), window.size = 1000000, border = 1)
kpPlotDensity(kp, data=gr_c, col = alpha("red", 0.6), window.size = 1000000, border = 1)
```
```{r Karyoplot3, echo = FALSE}
gr_tst <- gr[gr$HET == TRUE & gr$SNV == TRUE & gr$OVERLAP == FALSE & seqnames(gr) == "chr1"]

pp <- getDefaultPlotParams(plot.type=1)
pp$topmargin <- 15
kp <- plotKaryotype(plot.type = 1, 
                    chromosomes = c("chr1"), 
                    main = "QUAL scores Black is known, red is novel SNV",
                    plot.params = pp
)
kp <- kpDataBackground(kp)
kp <- kpAxis(kp, ymin = 0, ymax = 1, )
kp <- kpPoints(kp, chr = "chr1", 
               x = start(gr_tst), 
               y = gr_tst$QUAL/max(gr_tst$QUAL)) #All
kp <- kpPoints(kp, chr = "chr1", 
               x = start(gr_tst[gr_tst$NEW == T]), 
               y = gr_tst$QUAL[gr_tst$NEW == T]/max(gr_tst$QUAL),
               col = "red") #Only NEW

gr_tst <- gr[gr$HET == TRUE & gr$SNV == TRUE & gr$OVERLAP == FALSE 
             #& seqnames(gr) == "chr1"
             ]

pp <- getDefaultPlotParams(plot.type=4)
pp$topmargin <- 15
pp$ideogramheight <- 0
pp$data1inmargin <- 0
pp$data2inmargin <- 0
pp$leftmargin <- 0.2
kp <- plotKaryotype(plot.type = 4, 
                    chromosomes = paste0("chr", c(1:22, "X")), 
                    main = "Heterozygote/SNV/No_Overlap",
                    plot.params = pp,
                    labels.plotter = NULL,
                    ideogram.plotter = NULL
)
kp <- kpAddLabels(kp, labels = "QUAL Score", srt = 90)
kp <- kpDataBackground(kp)
kp <- kpAxis(kp, ymin = 0, ymax = max(gr_tst$QUAL)/10)
kp <- kpAddChromosomeNames(kp, srt = 90)
kp <- kpAddLabels(kp, labels = paste0("All (n = ", length(gr_tst), ")"), col = "1", r0 = 0.75, label.margin = 0.01, srt = 45, cex = 2)
kp <- kpAddLabels(kp, labels = paste0("New (n = ", length(gr_tst[gr_tst$NEW == T]),")"), col = "2", r0 = 0.25, label.margin = 0.01, srt = 45, cex = 2)
kp <- kpPoints(kp, 
               data = gr_tst,
               #chr = "chr1", 
               #x = start(gr_tst), 
               y = gr_tst$QUAL/max(gr_tst$QUAL)*10,
               col = alpha(colour = 1, alpha = 0.01)) #All
kp <- kpPoints(kp, 
               data = gr_tst[gr_tst$NEW == T],
               #chr = "chr1", 
               #x = start(gr_tst[gr_tst$NEW == T]), 
               y = gr_tst$QUAL[gr_tst$NEW == T]/max(gr_tst$QUAL)*10,
               col = alpha(colour = 2, alpha = 0.03)) #Only NEW

```






## Rsession
```{r session}
sessionInfo()
```
============